### PR TITLE
Make makeutype and combiners.h slightly less broken

### PR DESCRIPTION
**Pull request by jtanx**
Well this was a big rabbit hole. Initially I was just fixing compiler warnings on makeutype, but then I found I got a different result to what's currently committed for `utype.c` (see comment from #3755).

I ended up digging deeper, and it's actually because sometimes, the size of the lookup arrays in combiners.h didn't match the length of the ranges specified, so where the range is longer than the array, the generated output was garbage.

I've attempted to fix this by parsing appropriate blocks from UnicodeData.txt with https://gist.github.com/jtanx/bfa7489acc1af53f7937047908ca7704, and trying to follow the mapping that existed in `combiners.h` from [canonical combining classes](http://www.unicode.org/reports/tr44/#Canonical_Combining_Class_Values) to FF's internal representation.

But I'm no expert in this field, and to me, `combiners.h` is really broken. I'm probably missing something, but there are literally blocks that are specified that have no canonical combining class marked in UnicodeData.txt, so I have no idea how that was originally inferred.

 

### Type of change
- **Bug fix**

### CODE REVIEW

**Comment by JoesCat** [comment]
One rabbit hole you'll want to note is the different results you get between your:
```c
parts = line.strip().split(';')
```
and the scanf used in makeutype.c, I'm pretty sure it introduces errors as I found that out trying to make a better mirror().
You may want to run makecombiners.h against unicode #3 since that seems closer to original date, but if I recall, I think you might be closer to getting same results by doing a visual check against the pdf files.
Will look at the code tomorrow.

**Comment by jtanx** [comment]
> and the scanf used in makeutype.c, 

What is this with reference to? makeutype doesn't even use scanf.

Best I can tell, combiners.h was hand-curated; that's where I was fixing up the discrepancies. Functionally, makeutype is the same as before. The only changes in makeutype are mostly to just get rid of compiler warnings, plus the added assertion to make sure that the size of the lookup table from combiners.h matches the range defined.

**Comment by ctrlcctrlv** [comment]
Hand-curation seems likely to me. I've worked on this area of the code and had the same question @jtanx and that was my answer for it.

**Comment by JoesCat** [comment]
> I ended up digging deeper, and it's actually because sometimes, the size of the lookup arrays in combiners.h didn't match the length of the ranges specified, so where the range is longer than the array, the generated output was garbage.

is_Ligatures used whatever data was available in the arrays, but had to keep it's own data since some information like Arabic had many alternate characters. I avoided churning the waters too much by having is_ligatures done before digging into the code afterwards.

I concur, combiners looks hand-curated for the early code. the jump from 3->9 included visual inspection of the pdfs.

my error, not scanf, it's going to be these strtol functions, we need to split the lines at ";" and run these instructions on the separate strings. I started seeing errors while looking for mirror codes.
	alt = strtol(apt,&end,16)

**Comment by ctrlcctrlv** [comment]
The problem is that whatever changes that are being done based on inspection of Unicode charts (which, by the way, are not infallible&hellip;neither are Unicode's character properties, I know of several cases where they are wrong, off the top of my hand, ARMENIAN ABBREVIATION MARK (_pativ_) should be a combining mark, NOT a spacing mark) are not being marked as such so they can easily be carried over into new versions of Unicode.

I propose a rule: No hand-curation without separation. Meaning, hand-curated characters should all be together at the bottom of the file...e.g. `poses300[2] = FF_UNICODE_Above`

**Comment by JoesCat** [comment]
> I propose a rule: No hand-curation without separation. Meaning, hand-curated characters should all be together at the bottom of the file...e.g. ```poses300[2] = FF_UNICODE_Above```

It's been like this for a while: [Initial revision in 2001apr24](https://github.com/fontforge/fontforge/commit/2a768f98efe49c66a26b54308e298bbb2cf7807b#diff-18431dee3478ac45e2de1766f75ca8da)

File: Unicode/combiners.h [hunk 1]
```diff
@@ -697,7 +713,8 @@ static const int posesEC8[] = {
 };
 
 static const int posesF18[] = {
-    FF_UNICODE_Below|FF_UNICODE_Right
+    FF_UNICODE_Below,
```
[the remark this answers is not in the document]
Comment (jtanx) [reply Unicode/combiners.h]: ¯\\_(ツ)_/¯ I'm just following what UnicodeData.txt defines, if we want to make subjective changes to that, we should probably follow @ctrlcctrlv's suggestion

**Pull request merged**